### PR TITLE
Refactor GPT-OSS cache implementation

### PR DESCRIPTION
**Pull request by smedhe**
## Summary

  This PR refactors the GPT-OSS KV cache implementation to use a `QEffDynamicCache`-style per-layer cache instead of the previous list-backed `QEffHybridCacheForGPTOSS`.

  Key changes:
  - Preserves legacy tuple-cache compatibility through `from_legacy_cache()` and existing cache update entry points.
  - Updates GPT-OSS model code to use the new cache class and layer storage layout.
  - Enables GPT-OSS dynamo export / ORT parity and QAIC continuous batching tests by removing previous `xfail` / `skip` guards.
  - Updates cache correctness tests to cover the new GPT-OSS dynamic cache behavior, including chunked updates, sliding-window updates, legacy conversion, layer independence, and cache length tracking.

### CODE REVIEW

File: tests/unit_test/models/test_sliding_window_cache.py [hunk 1]
```diff
@@ -407,7 +425,7 @@ def test_full_cache_update_chunked_scatter_at_correct_position(self):
         k = torch.ones(batch, heads, 4, head_dim) * 5.0
         v = torch.ones(batch, heads, 4, head_dim) * 5.0
         k_out, v_out = cache.full_cache_update_chunked(
-            k, v, layer_idx=0, cache_kwargs={"position_ids": pos_ids(seq=4), "batch_index": None}
+            k, v, layer_idx=1, cache_kwargs={"position_ids": pos_ids(seq=4), "batch_index": None}
```
Comment (ochougul) [inline tests/unit_test/models/test_sliding_window_cache.py:428]: why is this changed?

File: QEfficient/transformers/cache_utils.py [hunk 2]
```diff
@@ -1461,6 +1396,126 @@ def sliding_window_update_chunked(
...
+    def __init__(
+        self,
+        config=None,
+        ddp_cache_data: Optional[Iterable[tuple[torch.Tensor, torch.Tensor]]] = None,
+        *args,
+        **kwargs,
+    ):
+        self.config = config
+        self._max_cache_len = None
+        self._sliding_window_len = None
```
Comment (ochougul) [inline QEfficient/transformers/cache_utils.py]: Remove not needed, you are already passing self.config and using that for sliding_window

**Review by ochougul — CHANGES_REQUESTED** [review]
can you please make sure that we have the tests for checking accuracy of following ways for gpt-oss with subfunc with dynamo for 2 layer tiny model
- prefill_only + enable_chunking  + expert_parallel moe
- decode-only + retain_full_kv

File: tests/unit_test/models/test_sliding_window_cache.py (hunk 1 above)
Comment (smedhe) [reply tests/unit_test/models/test_sliding_window_cache.py:428]: in gpt oss, layer 0 is sliding window and layer 1 is full cache, that's why for test_full_cache_update using layer idx 1

File: QEfficient/transformers/cache_utils.py (hunk 2 above)
Comment (smedhe) [reply QEfficient/transformers/cache_utils.py]: in the modeling file we use past_key_values.sliding_window_len, to maintain that I have added this. We derive the max cache len and sliding_window_len from the key shapes in the old code, doing the same here.

Comment (smedhe) [reply QEfficient/transformers/cache_utils.py]: will be needed for max_cache_len

**Comment by quic-amitraj** [comment]
CI-Ready

**Comment by ochougul** [comment]
CI-Ready

**Comment by smedhe** [comment]
CI-Ready

**Comment by smedhe** [comment]
CI-Ready

**Comment by quic-amitraj** [comment]
CI-Ready

**Comment by quic-amitraj** [comment]
CI-Ready

File: QEfficient/transformers/models/modeling_auto.py [hunk 3]
```diff
@@ -3448,6 +3448,22 @@ class QEFFAutoModelForCausalLM(QEFFBaseModel):
 
     _onnx_transforms = []
 
+    def _override_gptoss_prefill_chunking(
```
Comment (quic-amitraj) [inline QEfficient/transformers/models/modeling_auto.py]: Not good place to keep specific model related changes here

**Review by quic-amitraj — CHANGES_REQUESTED** [review]

File: examples/disagg_serving/gpt_oss_disagg_mode.py [hunk 4]
```diff
@@ -1,137 +0,0 @@
-# -----------------------------------------------------------------------------
```
Comment (quic-amitraj) [inline examples/disagg_serving/gpt_oss_disagg_mode.py:1]: Why deleted?

**Review by quic-amitraj — CHANGES_REQUESTED** [review]

File: examples/disagg_serving/gpt_oss_disagg_mode.py (hunk 4 above)
Comment (smedhe) [reply examples/disagg_serving/gpt_oss_disagg_mode.py:1]: We are disabling the support for prefill_only=True and enable_chunking=False, this was the example script for that. If the user passes this config, we issue a warning and override enable_chunking=True

File: QEfficient/transformers/models/gpt_oss/modeling_gpt_oss.py [hunk 5]
```diff
@@ -887,9 +888,7 @@ def forward(
 
         if output_attentions:
             outputs += (self_attn_weights,)
-
-        if use_cache:
-            outputs += (present_key_value,)
+        # not returning present_key_value as it is not used
```
Comment (ochougul) [inline QEfficient/transformers/models/gpt_oss/modeling_gpt_oss.py]: this comment is wrong, `we are not returning it because it is updated in-place` ?

File: QEfficient/transformers/models/gpt_oss/modeling_gpt_oss.py [hunk 6]
```diff
@@ -929,7 +928,7 @@ def forward(
         return_legacy_cache = False
         if use_cache and not isinstance(past_key_values, Cache):
             return_legacy_cache = True
-            past_key_values = QEffHybridCacheForGPTOSS.from_legacy_cache(self.config, past_key_values)
+            past_key_values = QEffGPTOSSDynamicCache.from_legacy_cache(self.config, past_key_values)
```
Comment (ochougul) [inline QEfficient/transformers/models/gpt_oss/modeling_gpt_oss.py]: why are we changing the name to `DynamicCache` ?

File: QEfficient/transformers/models/pytorch_transforms.py [hunk 7]
```diff
@@ -1622,12 +1622,6 @@ def apply(
             raise ValueError("moe expert_parallel_chunk_size must be greater than zero")
         compile_seq_len = prefill_seq_len or ONNX_EXPORT_EXAMPLE_SEQ_LEN
         num_packed_chunks = max(1, -(-compile_seq_len // expert_parallel_chunk_size))
-        if compile_seq_len % expert_parallel_chunk_size != 0:
-            logger.warning(
-                f"qaic_config['moe_config']['expert_parallel_chunk_size']={expert_parallel_chunk_size} does not evenly divide "
-                f"the compile sequence length {compile_seq_len}; the number of packed chunks will be {num_packed_chunks}."
-            )
-
```
Comment (ochougul) [inline QEfficient/transformers/models/pytorch_transforms.py:1630]: why is this moved?

**Review by ochougul — APPROVED** [review]
